fix(lines-wordwrap): support wordbreak and hyphens option on lines method

arnaud-zg opened the pull request:
Details on issue #2

bezoerb commented:
Thanks @arnaud-zg for working on this :)
I'll take a deeper look at the PR tomorrow.

bezoerb commented on src/utils.js:
```diff
@@ -417,7 +417,30 @@ export function computeLinesDefault({ctx, text, max, wordSpacing, letterSpacing}
     if ([...line].length !== 0) {
-        lines.push(line);
+        if (options.wordwrap) {
+            line
```
The wordwrap option should be considered above before the breakpoint switch...case. You also should only check the last part and not the entire line. (Think about a line with 1000 characters and only the last 10char word is to long)

arnaud-zg replied:
Right, I moved it just before the process loop over text parts and compute the lines, and add a breakpoint of type `SHY` for each new line.

arnaud-zg commented:
It's been a while that my pull request is on pending status, actually I need this feature on my project. It would be great if it can be merged this week.

bezoerb commented:
@arnaud-zg: sorry for taking so long to come back. This implementation may compute wrong values as the browser would not break the words like this (won't add `-` at the end) 
I think it's important that the result reflects the behavior from the browser. Did you know that you can use the `word-break: 'break-all'` css rule (either in css or overwrite in your call)
```js
 textMetrics(el).lines(text, {}, {'word-break': 'break-all'})
/*
[ 'Craspharetraph',
  'aretragravida.Vi',
  'vamusconsequ',
  'atlacusvelposu',
  'erecongue.Duis',
  'aloremvitaeexa',
  'uctorscelerisqu',
  'enoneuturpis.Ut',
  'imperdietmagn',
  'asitametjustobi',
  'bendumvehicul',
  'a.' ]
*/
```
Browser:
![image](https://user-images.githubusercontent.com/1697800/33800371-0a0ead66-dd3f-11e7-9416-aec3bf0b7625.png)

Sorry to not mention this before.
Does this help with your project or do you depend on `text-metrics` adding the `&shy;` to each line? 
You should also keep in mind, that adding the separator will at the end of each line will result in grammatical errors

arnaud-zg commented:
According to `word-break` and `hyphens` documentation, I updated my code and added those two option.
Case 1:
```
textMetrics(el).lines(text, {}, {wordBreak: 'break-all'})
/*
[ 'Craspharetraph',
  'aretragravida.Vi',
  'vamusconsequ',
  'atlacusvelposu',
  'erecongue.Duis',
  'aloremvitaeexa',
  'uctorscelerisqu',
  'enoneuturpis.Ut',
  'imperdietmagn',
  'asitametjustobi',
  'bendumvehicul',
  'a.' ]
*/```

Case 2:
```
textMetrics(el).lines(text, {}, {hyphens: 'auto'})
/*
[ 'Craspharetraph-',
  'aretragravida.Vi-',
  'vamusconsequ-',
  'atlacusvelposu-',
  'erecongue.Duis-',
  'aloremvitaeexa-',
  'uctorscelerisqu-',
  'enoneuturpis.Ut-',
  'imperdietmagn-',
  'asitametjustobi-',
  'bendumvehicul-',
  'a.' ]
*/
```

arnaud-zg commented:
Up

bezoerb commented on src/utils.js:
```diff
@@ -351,7 +351,37 @@ export function computeLinesDefault({ctx, text, max, wordSpacing, letterSpacing}
+        let breakPointType = null;
+
+        if (options.wordBreak === 'break-all') {
+            breakPointType = 'BAI';
```
As i said, the break-all case is already handled here: 
https://github.com/bezoerb/text-metrics/blob/master/src/index.js#L143
https://github.com/bezoerb/text-metrics/blob/master/src/utils.js#L426

bezoerb commented on test/test.js:
```diff
@@ -42,6 +42,60 @@ test('Computes maxFontSize', t => {
+        'agnasitametjustob-',
+        'bibendumvehicul-',
+        'la.'
+    ];
```
Hyphens 'auto' behaves different. If you want to add a hyphen at the end of each line you can do
```js
const value = textMetrics(el).lines(text.replace(/(.{1})/g,"$1&shy;"))
```

bezoerb commented:
Sorry @arnaud-zg, i don't think it's a good idea to add this feature as this won't match the expected behavior. When doing auto hyphens this should be done properly with grammatical correct results.

arnaud-zg commented:
:+1:
